updates for module composition (guidance for not using non-AVM modules) & minor doc updates (#133)
Azure/Azure-Verified-Modules: docs/content/specs/shared/_index.md

<br>

#### ID: RMFR8 - Category: Composition - Dependency on child and other resources

> A resource module **MAY** contain references to other resource modules, however **MUST NOT** contain references to non-AVM modules nor AVM pattern modules.

Reviewer: Can we add a call out to:

- https://azure.github.io/Azure-Verified-Modules/specs/bicep/#id-bcpfr1---category-composition---cross-referencing-modules
- https://azure.github.io/Azure-Verified-Modules/specs/terraform/#id-tffr1---category-composition---cross-referencing-modules

Author: Done

A Pattern Module **MAY** be exempt of this requirement if using Resource Modules would result in hitting scaling limitations and/or would reduce the capabilities of the Pattern Module due to the limitations of Azure Resource Manager. In this scenario, a Pattern Module **MAY** use "vanilla" code for those resources where it's necessary.

{{< hint type=note >}}
The intent behind this being a "**SHOULD**" level requirement (and not a **MUST**) is to cater for those scenarios where most Resource Modules required to develop a given Pattern Module are available, but not all. In this case, the AVM team wants to unblock the owner of the new Pattern Module by allowing them to leverage this approach as a workaround, until the required Resource Module becomes available. Ideally, all required Resource Modules **SHOULD** be developed first, and then leveraged by the Pattern Module.

> {{< /hint >}}

Author: The risk of accumulating technical debt is now called out specifically.

#### ID: PMNFR2 - Category: Composition - Use Resource Modules to Build a Pattern Module

> A Pattern Module **SHOULD** be built from the required AVM Resources Modules. If a Resource Module does not exist for a resource required by the Pattern Module, then the Pattern Module **MAY** use "vanilla" code for this resource, however, the Pattern Module **SHOULD** be updated to use the Resource Module when the required Resource Module becomes available.

Reviewer 2: ```suggestion
A Pattern Module **SHOULD** be built from AVM Resources Modules to establish a standardized code base & improve maintainability. If a Resource Module does not exist for a resource required by the Pattern Module, then the Pattern Module **MAY** use "vanilla" code for this resource, however, the Pattern Module **SHOULD** be updated to use the Resource Module when the required Resource Module becomes available.
```

> A Pattern Module **MAY** be exempt of this requirement if using Resource Modules would result in hitting scaling limitations and/or would reduce the capabilities of the Pattern Module due to the limitations of Azure Resource Manager. In this scenario, a Pattern Module **MAY** use "vanilla" code for those resources where it's necessary.

Reviewer 2: ```suggestion
A Pattern Module **MAY** opt not following this recommendation if using Resource Modules would result in hitting scaling limitations and/or would reduce the capabilities of the Pattern Module due to the limitations of Azure Resource Manager. In this scenario, a Pattern Module **MAY** use "vanilla" code for those resources where it's necessary.
```

{{< hint type=note >}}
The intent behind this being a "**SHOULD**" level requirement (and not a **MUST**) is to cater for those scenarios where most Resource Modules required to develop a given Pattern Module are available, but not all. In this case, the AVM team wants to unblock the owner of the new Pattern Module by allowing them to leverage this approach as a workaround, until the required Resource Module becomes available. Ideally, all required Resource Modules **SHOULD** be developed first, and then leveraged by the Pattern Module.

> Not adhering to this guidance would result in accumulating technical debt, as the Pattern Module **SHOULD**  be retrospectively updated to use the Resource Module once it becomes available.

Reviewer 2: ```suggestion
Not adhering to this guidance would result in accumulating technical debt and Pattern Modules **SHOULD** be retrospectively updated to use Resource Modules once they become available.
```

> Not adhering to this guidance would result in accumulating technical debt, as the Pattern Module would need to be retrospectively updated to use the Resource Module once it becomes available.

Reviewer: ```suggestion
Not adhering to this guidance would result in accumulating technical debt, as the Pattern Module **SHOULD**  be retrospectively updated to use the Resource Module once it becomes available.
```
